kv: bypass lease transfer safety checks during joint consensus

nvb opened the pull request:
This commit adds logic to bypass lease transfer safety checks (added in 034611b) when in a joint configuration and transferring the lease from a VOTER_DEMOTING to a VOTER_INCOMING. We do so because we could get stuck without a path to exit the joint configuration if we rejected this lease transfer while waiting to confirm that the target is up-to-date on its log. That confirmation may never arrive if the target is dead or partitioned away, and while we'd rather not transfer the lease to a dead node, at least we have a mechanism to recovery from that state. We also just sent the VOTER_INCOMING a snapshot (as a LEARNER, before promotion), so it is unlikely that the replica is actually dead or behind on its log.

A better alternative here would be to introduce a mechanism to choose an alternate lease transfer target after some amount of time, if the lease transfer to the VOTER_INCOMING cannot be confirmed to be safe. We may do this in the future, but given the proximity to the release and given that this matches the behavior in v22.1, we choose this approach for now.

Release note: None

Release justification: Needed to resolve release blocker.

Fixes: #88667
See also https://github.com/cockroachdb/cockroach/pull/89564

nvb committed:
kv: bypass lease transfer safety checks during joint consensus

Fixes #88667.

This commit adds logic to bypass lease transfer safety checks (added in
034611b) when in a joint configuration and transferring the lease from a
VOTER_DEMOTING to a VOTER_INCOMING. We do so because we could get stuck
without a path to exit the joint configuration if we rejected this lease
transfer while waiting to confirm that the target is up-to-date on its
log. That confirmation may never arrive if the target is dead or
partitioned away, and while we'd rather not transfer the lease to a dead
node, at least we have a mechanism to recovery from that state. We also
just sent the VOTER_INCOMING a snapshot (as a LEARNER, before
promotion), so it is unlikely that the replica is actually dead or
behind on its log.

A better alternative here would be to introduce a mechanism to choose an
alternate lease transfer target after some amount of time, if the lease
transfer to the VOTER_INCOMING cannot be confirmed to be safe. We may do
this in the future, but given the proximity to the release and given that
this matches the behavior in v22.1, we choose this approach for now.

Release note: None

Release justification: Needed to resolve release blocker.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/89340)
<!-- Reviewable:end -->

aayushshah15 commented:
Do you think we should add a unit test that asserts that the lease transfer to a dead VOTER_INCOMING works?





<!-- Sent from Reviewable.io -->

nvb committed:
kv: improve logging around maybeTransferLeaseDuringLeaveJoint

Also simplify a bit of logic.

nvb committed:
kv: add TestTransferLeaseDuringJointConfigWithDeadIncomingVoter

This commit adds a new test which ensures that the lease transfer
performed during a joint config replication change that is replacing the
existing leaseholder does not get stuck even if the existing leaseholder
cannot prove that the incoming leaseholder is caught up on its log. It
does so by killing the incoming leaseholder before it receives the lease
and ensuring that the range is able to exit the joint configuration.

Currently, the range exits by bypassing safety checks during the lease
transfer, sending the lease to the dead incoming voter, letting the
lease expire, acquiring the lease on one of the non-demoting voters, and
exiting. The details here may change in the future, but the goal of this
test will not.

nvanbenschoten commented:
Yes, an end-to-end unit test was needed here. We were already testing that the new flag was properly set in a few tests, but not that this allowed the range to exit a joint configuration when the VOTER_INCOMING is dead.

I've added a test in a new commit. It's flaky now without https://github.com/cockroachdb/cockroach/pull/89221 because the VOTER_DEMOTING is allowed to re-acquire the lease (the starvation we hypothesized), which @shralex is working on. We'll need to sequence these changes appropriately.

nvanbenschoten commented:
TFTR! I'll have to wait for https://github.com/cockroachdb/cockroach/pull/89564 to merge before this will pass CI.

nvanbenschoten commented:
bors r=shralex

shermanCRL commented:
Pardon if this was already discussed, should this backport to v22.1?

nvanbenschoten commented:
These lease transfer safety checks were only added in v22.2 (in 034611b), so this doesn't need a backport. We were hopeful in https://github.com/cockroachdb/cockroach/pull/82758 that the change would be non-disruptive and eventually qualify for a backport, but as with many changes in this area, the effects are far-reaching, so we did not gain enough confidence to meet the backport criteria.
